### DISCUSSION TITLE
Update jiminy 1.8.3 -> 1.8.5.post1, extreme slow down using "spring_damper" model

### DISCUSSION THREAD

**Discussion by mwulfman** [Wandercraft/jiminy#785] [Q&A]
I have just tried updating to the last version of jiminy (1.8.5.post1) from jiminy 1.8.3 and while the "spring_damper" model was already fairly slow for my use case, now it is significantly slower for some simulations. Any ideas where that could come from, or if I need to adapt solver parameters following the update to solve this issue ?

**Comment by duburcqa** [comment]
Interesting... Do you known if `1.8.4` was causing the same issue (and ideally also `1.8.5`). To my knowledge I changed nothing regarding the spring-damper model in particular. May be it is "just" `reset that is slower than before, slowing down the whole process. Now we are logging the whole robot. But I don't expect that to take so much time since it is pure C++.

**Reply by mwulfman** [reply]
To give more context and answer your question. On the same machine, with spring_damper on:
On `1.8.3`, I am able to run 10 simulations of 30s in ~4.8s. 
On `1.8.4`, I am able to run 10 simulations of 30s in ~200s. (I've set `debug=True` otherwise it timeouts every time)
I had to change my code to switch to >`1.8.4` so maybe I've created a bug but I believe I've just done the API changes related to the way the options are handled now.

Just resetting the environment:
On `1.8.3`, resetting the simulation environment 10 times takes ~0.28s.
On `1.8.4`, resetting the simulation environment 10 times takes ~0.18s.

Reset + 1 step, 10 times, in the environment:
On `1.8.3`, ~0.42s.
On `1.8.4`, ~14s.

Reset + 2 steps, 10 times, in the environment:
On `1.8.3`, ~0.6s.
On `1.8.4`, ~30s.

**Reply by duburcqa** [reply]
OK there is definitely a big issue. Taking so long for doing a single step makes no sense. You should run `py-spy` and upload the flame chart here so that we could investigate the bottleneck. But I doubt it is coming from jiminy in light of the magnitude of the slow down.

**Reply by mwulfman** [reply]
To provide more infos while I conduct more investigation. With the "constraint" model:
On `1.8.3`, 100 simulations of random action of 30s max take ~7s
On `1.8.4`, 100 simulations of random action of 30s max take ~6s. (accompagnied with the warning
```python
WARNING:jiminy::Engine::computeAcceleration(/__w/jiminy/jiminy/core/src/engine/engine.cc:3886):
Constraint solver failure.
```
)

So I see an improvement with the other contact model.

py-spy flame graph only shows that "step (command/envs/generic.py:930)" takes 98% of the CPU time.

**Reply by duburcqa** [reply]
Yes, I'm not so surprise to see some improvements with the other contact model even though I did not work on it directly. Regarding your slow-down. I suspect it is related to the new backlash modelling. There is no virtual inertia anymore, which is probably necessary to stabilise the dynamics when using "spring_damper" contact model. Are you using them ? If so, you should add virtual inertia on them manually.

**Reply by duburcqa** [reply]
The contact solver failure message is not a big deal. This is supposed to happen from time to time. But if you start to be flooded, only then there would be an issue.

**Reply by duburcqa** [reply]
If backlashes are causing the issue, then I'm not going to "fix" it. The plan is rather to get rid of the old 'spring_damper' contact model because it is numerically unstable to the point to be barely usable in many cases. 

Short term, I think I will force using the constraint model for joint bounds, and only take into account the user-specified contact model when it comes to ground contact interactions. Solving joint bounds with the constraint solver should be fast enough anyway, since I'm not expecting all bounds to be active at once, and anyway each constraint is 1DoF.

**Reply by mwulfman** [reply]
> Yes, I'm not so surprise to see some improvements with the other contact model even though I did not work on it directly. Regarding your slow-down. I suspect it is related to the new backlash modelling. There is no virtual inertia anymore, which is probably necessary to stabilise the dynamics when using "spring_damper" contact model. Are you using them ? If so, you should add virtual inertia on them manually.

I was thinking I was not using them (setting enableBacklash=False at initialisation), but there must be a bug in my code because they are activated in my robot's options... thanks for your help, I will try that!

**Reply by mwulfman** [reply]
Yes, the enableBacklash option was getting overridden at reset...

**Reply by duburcqa** [reply]
What do you mean ? You were overriding it or it is a bug in Jiminy ?

**Reply by mwulfman** [reply]
I was overriding it. With backlashes disabled, I ran 100 simulation in 26s. Thanks !

**Reply by duburcqa** [reply]
So it is still much slower than before ?

> On 1.8.3, I am able to run 10 simulations of 30s in ~4.8s.

**Reply by mwulfman** [reply]
No, it appears faster 4.8s versus 2.6s for 10 simulations. Though it is simulations with heterogeneous duration (in which we give random position command) so further testing is needed to confirm the speed increase.

**Reply by duburcqa** [reply]
Ah yes indeed i missed one zero. Glad to hear that all sortef out then !
